### PR TITLE
fix a bug: in value.h, ValueConstIterator can convert to ValueIterato…

### PR DESCRIPTION
**Pull request by ycqiu**
…r, I think that is a bug. the correct way is ValueIterator can convert to ValueConstIterator.

### CODE REVIEW

**Comment by BillyDonahue** [comment]
Totally. An implicit conversion allowing mutable access to const elements is the wrong direction. :)
I hope this gets fixed. Can you add a regression test for it?

**Comment by ycqiu** [comment]
#include <json/json.h>
#include <iostream>
#include <sstream>
#include <string>

using namespace std;


int main()
{
	Json::Value value;

	for(int i = 10; i < 20; ++i)
	{
		stringstream sstream;	
		sstream << i;
		string str;
		sstream >> str;
		//cout << str << endl;
		value[str] = str;	
	}	

	//in old code, this will get a compile error
	Json::Value::const_iterator iter = value.begin();
	for(; iter != value.end(); ++iter)
	{
		cout << *iter << endl;
	}
	return 0;
}

**Comment by ycqiu** [comment]
yes. the test is easy, i send it here

#include <json/json.h>
//#include "../jsoncpp/include/json/json.h"
#include <iostream>
#include <sstream>
#include <string>

using namespace std;


int main()
{
	Json::Value value;

	for(int i = 10; i < 20; ++i)
	{
		stringstream sstream;	
		sstream << i;
		string str;
		sstream >> str;
		//cout << str << endl;
		value[str] = str;	
	}	

	//in old code, this will get a compile error
	Json::Value::const_iterator iter = value.begin();
	for(; iter != value.end(); ++iter)
	{
		cout << *iter << endl;
	}
	return 0;
}

**Comment by ycqiu** [comment]
yes. the test is easy, i send it here.  the include file can not put here.

int main()
{
Json::Value value;

for(int i = 10; i < 20; ++i)
{
    stringstream sstream;   
    sstream << i;
    string str;
    sstream >> str;
    //cout << str << endl;
    value[str] = str;   
}   

//in old code, this will get a compile error
Json::Value::const_iterator iter = value.begin();
for(; iter != value.end(); ++iter)
{
    cout << *iter << endl;
}
return 0;
}

**Comment by ycqiu** [comment]
yes. I send code here. the include file can not put here.


int main()
{
	Json::Value value;

	for(int i = 10; i < 20; ++i)
	{
		stringstream sstream;	
		sstream << i;
		string str;
		sstream >> str;
		//cout << str << endl;
		value[str] = str;	
	}	

	//in old code, this will get a compile error
	Json::Value::const_iterator iter = value.begin();
	for(; iter != value.end(); ++iter)
	{
		cout << *iter << endl;
	}
	return 0;
}

**Comment by BillyDonahue** [comment]
Put the test in the PR. In comments it doesn't do much good. :)

**Comment by ycqiu** [comment]
ok, I puted the test.cpp in the PR, please check it.

File: test/test.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#include <json/json.h>
+//#include "../jsoncpp/include/json/json.h"
+#include <iostream>
+#include <sstream>
+#include <string>
+
+using namespace std;
```
Comment (BillyDonahue) [inline test/test.cpp]: Please remove this. It's a bad practice.

File: test/test.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+
+int main()
```
Comment (BillyDonahue) [inline test/test.cpp]: Can you just add this code to an existing test?

**Comment by ycqiu** [comment]
yes. 
sorry,this is my first time to use github.
I try it mybest.

**Comment by BillyGoto** [comment]
Sent from my iPhone

> On Oct 8, 2015, at 10:41, ycqiu <notifications@github.com> wrote:
> 
> yes. 
> sorry,this is my first time to use github.
> I try it mybest.
> 

Sure I just want to make this easier for you. Making a new test file means adding a build rule to build and run it. It's easier to piggyback on an existing target.

**Comment by cdunn2001** [comment]
Thanks for catching this, @ycqiu.

@BillyDonahue, I wonder: What if instead of removing the ctor, we include it but mark it `explicit`?

**Comment by BillyDonahue** [comment]
I don't think it's ever ok to make an `iterator` out of a `const_iterator`.
An explicit constructor could be triggered accidentally by template code or by an initializer and break const-correctness. It really should be removed.

**Comment by ycqiu** [comment]
i pull the code, please check it.

**Comment by cdunn2001** [comment]
You're right, of course. The conversion is a bad idea. But since it already existed, I had to keep it and just make it throw.

**Pull request closed without merging**